### ISSUE TITLE
Better QueryResult structure for go

### ISSUE THREAD

**Issue by dominichamon** [youtube/vitess#1226]
Currently, go clients work with `mproto.QueryResult` which contains `Field`s as a slice of structs with name, type, and flags. The rows, similarly, are a slice of slice of `sqltypes.Value`.

This leads to awkward code to work with results that requires looping through fields to pull out the name and then for each row, converting the values and handling them differently based on the name. It's possible to generally convert a result to:

```
type Result struct {
        InsertId uint64
        Rows     []map[string]interface{}
        Err      *RPCError
}
```

Note, the rows have already been converted, and are now indexed by field name. This is much nicer to work with, requiring only type assertions on the values.

I have a prototype that works like this performing the conversion and storing this new result in place of the `mproto.QueryResult` in `proto.QueryResult`, but I think it breaks abstraction boundaries as I expect everything should be working with `mproto` types.

Is it possible to change the struct definitions in `mproto` or has that ship sailed?
Would it be possible to migrate to a different version of the struct that is much nicer to work with as above?
Perhaps we could at least provide a utility method for taking an `mproto.QueryResult` and converting it to the above structure?

Speculation and discussion encouraged please!

**Comment by enisoc** [comment]
That sounds like where we want to go. One thing I should mention is I think were planning to get rid of mproto.QueryResult and replace it everywhere with the [QueryResult struct](https://github.com/youtube/vitess/blob/master/go/vt/proto/query/query.pb.go#L354) generated from our [protobuf](https://github.com/youtube/vitess/blob/master/proto/query.proto#L148). However, that struct is basically the same as mproto.QueryResult, so your point still stands.

In other languages, we've been wrapping QueryResult in a cursor abstraction that provides type conversion for rows. For example, see the Java [Row](https://github.com/youtube/vitess/blob/master/java/client/src/main/java/com/youtube/vitess/client/cursor/Row.java) class. It sounds like you're proposing something similar for the Go client, and I like that idea.

Can you elaborate more on why you worry this would break abstraction boundaries? Would you be in favor of the wrapper-type implementation we used in Java?

We also have a [Vitess driver](https://github.com/youtube/vitess/blob/master/go/vt/client/rows.go) for the [Go sql library](https://golang.org/pkg/database/sql/driver/) that does something similar to what you want. However, that driver only works with the [VTGate V3](http://vitess.io/doc/VTGateV3/) API, which is still in development.

**Comment by dominichamon** [comment]
Ah, I hadn't seen the new `QueryResult` struct. Yes, my thought is very similar to the Java Row class. I'm not averse to a wrapper, though it does add more boilerplate to the caller. I'd prefer if I just received a struct I can use immediately.

It's unfortunate that maps can't be generated from go protos as I think we'll end up with another conversion there, which is O(N) for the number of cells returned. At least in the current implementation the caller takes that hit once when they do the conversion and get the value.

My prototype (which I haven't pushed anywhere yet) ends up replacing `mproto.QueryResult` with something from `vtgate/proto` everywhere in the stack. I had a gut feeling that the `mproto` structs are meant to be the common types and the `vtgate` structs are an implementation detail for vtgate specifically. This is unique to my implementation though. There are others that wouldn't have this issue (for instance, adding the wrapper/new structs to mproto).

**Comment by enisoc** [comment]
Just the clarify, the Cursor wrapper in our new Java client is obligatory - the execute methods wrap QueryResult in Cursor before returning it to the caller. This has its own trade-offs, but I thought it was important to avoid boilerplate for the caller as you mentioned.

**Comment by sougou** [comment]
We also need to be network efficient. Sending the column name for every value is a pretty substantial cost.
Even sending the Fields once per result was expensive. I once wrote a prototype that cached the Fields on the client side, which achieved higher throughputs, but the code was too complicated.
I was basically competing with memcache :). There was no way to beat it, but we had to come close. Otherwise, the rowcache wasn't going to be viable.
So, we compromised on what you see today as mproto.QueryResult.

**Comment by michael-berlin** [comment]
Yes, Sugu, you're right about not repeating the field names over the wire. From what I've understood Dominic is more concerned about the API used by applications above Vitess.

First of all, let's start of with an overview of the different places where we touch any query responses:

1. mysql daemon: query results are returned back to us
2. vttablet: receive query results via the mysql C client library and store them in an internal data structure
3. vttablet: send out the internal data structure as RPCs (e.g. gRPC or BSON RPC)
4. vtgate: receive and forward the RPCs from vttablet to the client
5. client: at the RPC layer, receive the query result and store it in an internal data structure
6. client: through an API, hand-over the query result to the application in a suitable form (e.g. the internal data structure or a pre-processed form)

Normally, the internal data structure within vttablet and the client would be Google protobufs. But since we supported BSON RPC, we had to define our own data structures and couldn't use protobuf. That's the reason why there is an "mproto.QueryResult" struct. With the switch to gRPC, this structure will go away and all our code can use protobuf internally.

My guess is Dominic is fine with keeping things as they are for the internal data structure. He is more interested in making layer 6, the client API more user friendly. Currently, we have to do a conversion from the "raw" strings, which we receive from the MySQL client library and pass along in the protobuf, to the respective native types anyway. That's the point where Dominic suggests that we should return the query result indexed by field names.

Dominic, to answer your points:

> My prototype (which I haven't pushed anywhere yet) ends up replacing mproto.QueryResult with something from vtgate/proto everywhere in the stack.

This is our long term goal and we're working towards it. But we're not there yet and therefore we haven't dropped these internal "mproto" structures yet.

> It's unfortunate that maps can't be generated from go protos as I think we'll end up with another conversion there, which is O(N) for the number of cells returned. At least in the current implementation the caller takes that hit once when they do the conversion and get the value.

What do you mean here? Can you give an example?

With proto3, we can actually define maps in the protobuf specification and they translate very well to the expected Go code. But as said before by Sugu, we don't want to repeat field names in the wire format and also not when processing the data internally.

To sum things up: We can implement your proposal and I think it's a good idea. But only at the boundary between the Go client and the application. At the other layers, we prefer to keep things "raw" and unprocessed as possible.

**Comment by dominichamon** [comment]
Sounds good to me. I'm happy to work on that if someone can point me at the place I should do the conversion. Ie, should my changes be restricted to the vt/client directory?

**Comment by enisoc** [comment]
The `go/vt/client` package is the one that conforms to the `database/sql/driver` interface, so that one already converts to the native types expected by the interface.

The [vtgateconn](https://github.com/youtube/vitess/tree/master/go/vt/vtgate/vtgateconn) package is the one that would benefit from a type-converting result wrapper, in the style of our [Java Cursor](https://github.com/youtube/vitess/tree/master/java/client/src/main/java/com/youtube/vitess/client/cursor).

If you'd like to work on it, we should start by agreeing on some examples of exactly what it would look like to use the new interface. For example, it's not obvious how we'd want to wrap streaming results since those are usually channels in Go - whereas in Java we made StreamCursor use the same interface as a non-streaming Cursor.

**Comment by sougou** [comment]
This issue is somewhat related to #1620, which will allow us to build maps that many standard DB APIs prefer. I'll close this issue in favor of that one.
